remotespark/kernels/wrapperkernel/usercodeparser.py

```diff
@@ -0,0 +1,120 @@
+# Copyright (c) 2015  aggftw@gmail.com
+# Distributed under the terms of the Modified BSD License.
+
+from remotespark.utils.constants import Constants
+
+
+class UserCodeParser(object):
+    @staticmethod
+    def get_code_to_run(code, language):
+        code = code.strip()
+
+        lines = code.split("\n")
+
+        first_line_index = 0
+
+        if len(lines) == 0:
+            first_line = code
+        else:
+            first_line = lines[0].strip()
+            in_multi_line = False
+            start_used = None
+            for i in range(len(lines)):
+                line = lines[i].strip()
+
+                comment, in_multi_line, start_used = UserCodeParser._is_comment(language,
+                                                                                line,
+                                                                                in_multi_line,
+                                                                                start_used)
+                if not comment:
+                    first_line = line
+                    first_line_index = i
+                    break
+
+        if not first_line.startswith("%"):
+            code_to_run = "%%spark\n{}".format(code)
+        elif first_line.startswith("%local"):
+            code_to_run = UserCodeParser._code_to_run_helper(first_line, lines[first_line_index+1:], 6)
+        elif first_line.startswith("%%local"):
+            code_to_run = UserCodeParser._code_to_run_helper(first_line, lines[first_line_index+1:], 7)
+        else:
+            # First line contains a magic that is not %local
+            code_to_run = "\n".join(lines[first_line_index:])
+
+        return code_to_run
+
+    @staticmethod
+    def _code_to_run_helper(first_line, lines, l):
+        rest = "\n".join(lines)
+
+        if len(first_line) > l:
+            separator = "\n"
+            first_line_next = first_line[l+1:]
+        else:
+            separator = ""
+            first_line_next = first_line[l:]
+
+        return "{}{}{}".format(first_line_next, separator, rest)
+
+    @staticmethod
+    def _is_comment(language, line, in_multi_line, start_used):
+        """Returns (bool, bool, str) where:
```

Comment:
The API surrounding parsing comments here is pretty messy, and (though I haven't tested it) I'm fairly certain that it doesn't work in 100% of cases (example: in order to implement a proper multiline comment parser for Python, what you actually need is a string parser, and this code recognizes no escape sequences).  

In order to make everyone's lives easier what you instead want is the function:

    def _chomp_comment(language, code):
         """Given a block of code in the source language, return all the code after an initial comment if there is one."""
        ...

Then the implementation of `_chomp_comment` can simply iterate through all the characters in the code block and return a new string after all the comments/whitespace at the beginning.  This is more flexible and allows you to actually capture all the corner cases.